optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,415 @@ class _ParzenEstimatorParameters(
-            prior_mu = 0.5 * (low + high)
-            prior_sigma = 1.0 * (high - low)
-            if mus.size == 0:
-                low_sorted_mus_high = numpy.zeros(3)
-                sorted_mus = low_sorted_mus_high[1:-1]
-                sorted_mus[0] = prior_mu
-                sigma = numpy.asarray([prior_sigma])
-                prior_pos = 0
-                order = numpy.array([]).astype(int)
-            else:  # When mus.size is greater than 0.
-                # We decide the place of the  prior.
-                order = numpy.argsort(mus)
-                ordered_mus = mus[order]
-                prior_pos = int(numpy.searchsorted(ordered_mus, prior_mu))
-                # We decide the mus.
-                low_sorted_mus_high = numpy.zeros(len(mus) + 3)
-                sorted_mus = low_sorted_mus_high[1:-1]
-                sorted_mus[:prior_pos] = ordered_mus[:prior_pos]
-                sorted_mus[prior_pos] = prior_mu
-                sorted_mus[prior_pos + 1 :] = ordered_mus[prior_pos:]
+            shape = (n_observations + 1, len(choices))
+            assert prior_weight is not None
+            value = prior_weight / (n_observations + 1)
         else:
-            order = numpy.argsort(mus)
-            # We decide the mus.
-            low_sorted_mus_high = numpy.zeros(len(mus) + 2)
-            sorted_mus = low_sorted_mus_high[1:-1]
-            sorted_mus[:] = mus[order]
-
-        # We decide the sigma.
-        if mus.size > 0:
-            low_sorted_mus_high[-1] = high
-            low_sorted_mus_high[0] = low
-            sigma = numpy.maximum(
-                low_sorted_mus_high[1:-1] - low_sorted_mus_high[0:-2],
-                low_sorted_mus_high[2:] - low_sorted_mus_high[1:-1],
-            )
-            if not consider_endpoints and low_sorted_mus_high.size > 2:
-                sigma[0] = low_sorted_mus_high[2] - low_sorted_mus_high[1]
-                sigma[-1] = low_sorted_mus_high[-2] - low_sorted_mus_high[-3]
+            shape = (n_observations, len(choices))
+            assert prior_weight is not None
+            value = prior_weight / n_observations
+        weights = np.full(shape, fill_value=value)
+        weights[np.arange(n_observations), observations] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_numerical_params(
+        self, observations: np.ndarray, param_name: str
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        n_observations = self._n_observations
+        consider_prior = self._parameters.consider_prior
+        consider_magic_clip = self._parameters.consider_magic_clip
+        multivariate = self._parameters.multivariate
+        sigmas0 = self._sigmas0
+        low = self._low[param_name]
+        high = self._high[param_name]
+        assert low is not None
+        assert high is not None
+
+        if n_observations == 0:
+            consider_prior = True
+
+        pairs_of_observation_and_idx = np.asarray(
```

Comment:
I'm wondering if it's not both faster an easier to read if we create a new separate array for the indices and keep two separate ndarrays.

Reply:
Sounds good. I'll keep two separate ndarrays here.